### ISSUE TITLE
[Datasets] `split_at_indices` for `DatasetPipeline`

### ISSUE THREAD

**Issue by amogkam** [ray-project/ray#18127]
The `split_at_indices` feature implemented as part of https://github.com/ray-project/ray/issues/17592 looks great!

Could this be implemented for `DatasetPipeline` as well? This will allow us to do per-epoch global shuffling while also splitting the dataset to make sure each worker has the same amount of data (`ds.repeat().random_shuffle().split_at_indices(...)`) which is a requirement for distributed training.

**Comment by clarkzinzow** [comment]
@amogkam I think that we should add this for API coverage, although I don't think that this should block the train-test-validation split for distributed training. Please correct me if I'm wrong, but given that the test set is static and that shuffling the validation set is nonsensical (the loss/accuracy measure isn't sensitive to sample order), we'll want to create a pipeline off of the dataset _after_ the train-test-validation split, right? Otherwise we'd be doing unnecessary shuffling.

E.g., we'd want to do the following if we were going for an 80:10:10 train-test-validation split:
```python
@ray.remote(num_gpus=1)
class TrainingWorker:
    def train(self, train_shard_pipe: DatasetPipeline, validation_shard_pipe: DatasetPipeline):
        for epoch, (train_ds, validation_ds) in enumerate(
                zip(
                    train_shard_pipe.iter_datasets(),
                    validation_shard_pipe.iter_datasets())):
            # Train.
            for batch_idx, batch in enumerate(train_ds.to_torch(...)):
                # Process batch.

            # Calculate validation loss.
            for batch_idx, batch in enumerate(validation_ds.to_torch(...)):
                # Process batch.

            # Report validation loss.
            

ds = # some loaded dataset that has already been globally shuffled once
count = ds.count()

# 80:10:10 train-test-validation
train_ds, test_ds, validation_ds = ds.split_at_indices([int(0.8 * count), int(0.1 * count)])

# Create training set and validation set pipelines.
train_pipe = train_ds.repeat().random_shuffle()
validation_pipe = validation_ds.repeat()

# Instantiate all training workers.
workers = [TrainingWorker.remote() for _ in range(num_workers)]

# Split training set and validation set pipelines, giving a shard of each to each training worker.
train_runs = [
    worker.train.remote(train_shard_pipe, validation_shard_pipe)
    for worker, train_pipe_shard, validation_pipe_shard in zip(
        workers,
        train_pipe.split(len(workers), locality_hints=workers),
        validation_pipe.split(len(workers), locality_hints=workers),
    )]
```

**Comment by amogkam** [comment]
@clarkzinzow ah the splitting here isn't for a train-val-test split. It's for splitting the train dataset across multiple workers. So if I have N training workers, I want to split the dataset into N shards, with each shard having the same amount of data (the dataset will be padded so it'll be divisible by N). The reason we need `split_at_indices` instead of `split` for sharding the dataset is to make sure that each shard is the same size. `split` makes no guarantees here, and can end up in situations where Shard 0 contains the entire dataset and the rest of the shards don't contain anything.

**Comment by amogkam** [comment]
That's why it'll also be great if we could eventually add locality hints to `split_at_indices`.
